internal/v4/api.go

```diff
@@ -815,3 +826,44 @@ func (h *Handler) serveChangesPublished(_ http.Header, r *http.Request) (interfa
 func (h *Handler) serveMacaroon(_ http.Header, _ *http.Request) (interface{}, error) {
 	return h.newMacaroon()
 }
+
+// GET id/promulgate
+// TODO (mhilton): document admin/promulgate endpoint
+func (h *Handler) serveAdminPromulgate(id *charm.Reference, _ bool, w http.ResponseWriter, req *http.Request) error {
+	if err := h.authorize(req, []string{"promulgators"}); err != nil {
+		return errgo.Mask(err, errgo.Any)
+	}
+	if req.Method != "PUT" {
+		return errgo.WithCausef(nil, params.ErrMethodNotAllowed, "%s not allowed", req.Method)
+	}
+	data, err := ioutil.ReadAll(req.Body)
+	if err != nil {
+		return errgo.Mask(err)
+	}
+	var promulgate params.PromulgateRequest
+	if err := json.Unmarshal(data, &promulgate); err != nil {
+		return errgo.WithCausef(err, params.ErrBadRequest, "")
+	}
+	if promulgate.Promulgate {
+		if err := h.store.Promulgate(id); err != nil {
+			return errgo.Mask(err, errgo.Any)
+		}
+	} else {
+		if err := h.store.UpdateBaseEntity(id, bson.D{{"$set", bson.D{{"promulgated", mongodoc.IntBool(false)}}}}); err != nil {
```

Comment:
It does seem slightly odd that this isn't in charmstore alongside
Promulgate.
I wonder if we should have Store.SetPromulgated(id, isPromulgated).
The isPromulgated=true case is trivial, admittedly, but at least
that keeps the base entity promulgation code all in one place.